interfaces/builtin/serial_port.go

```diff
@@ -111,21 +174,47 @@ func (iface *SerialPortInterface) PermanentPlugSnippet(plug *interfaces.Plug, se
 func (iface *SerialPortInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
 	switch securitySystem {
 	case interfaces.SecurityAppArmor:
-		return []byte(fmt.Sprintf("%s rwk,\n", iface.path(slot))), nil
-	case interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		if len(slot.Attrs) == 1 {
+			// Path to fixed device node (no udev tagging)
+			path, pathOk := slot.Attrs["path"].(string)
+			if !pathOk {
+				return nil, nil
+			}
+			cleanedPath := filepath.Clean(path)
+			return []byte(fmt.Sprintf("%s rw,\n", cleanedPath)), nil
+		}
+
+		if len(slot.Attrs) == 3 {
+			// Wildcarded apparmor snippet as the cgroup will restrict down to the
+			// specific device
+			return []byte("/dev/serial-port-* rw,\n"), nil
+		}
+
+		return nil, nil
+	case interfaces.SecurityUDev:
+		usbVendor, vOk := slot.Attrs["usb-vendor"].(int)
+		if !vOk {
+			return nil, nil
+		}
+		usbProduct, pOk := slot.Attrs["usb-product"].(int)
+		if !pOk {
+			return nil, nil
+		}
+		var udevSnippet bytes.Buffer
+		udevSnippet.WriteString(udevHeader + "\n")
+		for appName := range plug.Apps {
+			udevSnippet.WriteString(fmt.Sprintf(udevDevicePrefix, usbVendor, usbProduct))
+			tag := fmt.Sprintf("snap_%s_%s", plug.Snap.Name(), appName)
+			udevSnippet.WriteString(fmt.Sprintf(udevTagSuffix, tag) + "\n")
```

Comment:
@niemeyer - a nice future improvement might be to move the tag generation out to a helper function for any interface to use so that each interface that needs it doesn't have to generate it itself. I am not blocking on this.